ambry-network/src/main/java/com.github.ambry.network/SocketNetworkClient.java

```diff
@@ -0,0 +1,416 @@
+/**
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.network;
+
+import com.github.ambry.clustermap.DataNodeId;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.config.NetworkConfig;
+import com.github.ambry.utils.Time;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A {@link NetworkClient} that provides a method for sending a list of requests in the form of {@link Send} to a host:port,
+ * and receive responses for sent requests. Requests that come in via {@link #sendAndPoll(List, Set, int)} call,
+ * that could not be immediately sent are queued, and an attempt will be made in subsequent invocations of the call (or
+ * until they time out).
+ * (Note: We will empirically determine whether, rather than queueing a request,
+ * a request should be failed if connections could not be checked out if pool limit for its hostPort has been reached
+ * and all connections to the hostPort are unavailable).
+ *
+ * This class is not thread safe.
+ */
+public class SocketNetworkClient implements NetworkClient {
```

Comment:
Note: code in this class copied from old NetworkClient class which is now an interface.